### ISSUE TITLE
creating large matched meshes from connectivity and coordinates

### ISSUE THREAD

**Issue by cwsmith** [SCOREC/core#205]
A large structured hex (eventually mixed) mesh is needed and there are issues getting a mesh generator to easily produce this.

The [`matchedReader` branch](https://github.com/cwsmith/core/tree/matchedReader) contains a driver named `matchedNodeElmReader` that calls the `apf::Construct` api to create an mds/pumi mesh from a files containing connectivity and coordinates. 

There are no new/special compilation flags needed to build this branch of core and the driver.

Run the driver with the following command:
```
./matchedNodeElmReader box_small.cnn box_small.crd NULL outModel.dmg outMesh.smb
```

`NULL` is the place holder argument for the file containing the matching info.

**Note, I am still testing this in parallel without matching.  I'll update this text and the issue when it is ready for your initial testing.**

**Comment by cwsmith** [comment]
Pasting Ken's email below

> Cameron and I met today to discuss the plan for this effort.  He is close to having something to share.
> 
> 
> Remove counter from the first column of all files and assume that line number is the global entity number (with obvious adjustment for multiple topology connectivity):
> 
> file 1  coordinates (forgot what we named it)
> Coordinates
> numVerts 3
> x1  x2  x3  
> (repeat for remainder (numVerts -1) times)
> 
> File 2  connectivity (again forgot what we called it)
> 
> numElmThisTop nElmNodes
> gn1 gn2 …. nElmNodes 
> (repeat for remainder (numElmThisTop -1) times)
> Repeat header and rectangular array for each topology.  Could put into different files initially if easier and then cat the files to create File 2.
> 
> 
> File 3 Matching
> numVerts 1
> -1 if unmatched
> Node number matched with
> 
> 
> File 4   Classification
> numVerts 1
> 0=fully interior of the volume
> 1-6 =classified on face (not edge or vertex)
> 11-22 = classified on model edge (not end points which are model vertices)
> 31-38 = classified on a model vertex.
> 
> I suppose on this last one we should agree on a master numbering.  Cameron, I know SCOREC used to have  canonical numbering for hex elements and if you can find that we should stick with it (it probably also exists in Simmetrix documentation too)?
> 
> File 5) Boundary conditions
> We should probably recycle our existing flat text smd or spj file format with the only change being that model tag is replaced by the classification tag described in File 4.
> 
> Riccardo mentioned in Slack that he has a 1B element case ready for testing and called it his largest so I suspect he has some in between which would be good.
> 
> 
> Current plan:
> 1) Cameron is going to push a version of his code that ignores matching  to github as soon as he confirms it works on the already provided 175k element on 1, 2, and 4 target mds parts
> 2) Riccardo and I will retest  1)
> 3) Riccardo and I will begin to test  progressively larger cases
> 4) Cameron will move on to add  features in the following order:
>          a) matching capability
>          b)  classification ingestion
>          c) Boundary condition ingestion and connection a), b) and c) such that Chef has what it would have gotten from smd and mesh previously. 
> 5) Riccardo and I test features as they get added and obviously run PHASTA to verify that all the information is coming through
> 
> 
> We are leaving Initial conditions out but Riccardo can correct me if I am wrong, we have already in the codes that he is using a way to overwrite good, spatially varying initial and boundary condition values.  The latter may or may require correct BC codes but I think the above will do that.
> 
> Let me know if I have forgotten anything or gotten any of if differently than we discussed today.
> 
> 
> Best regards,
> 
> Ken

**Comment by cwsmith** [comment]
The parallel reader now works.

**Comment by KennethEJansen** [comment]
Am I correct in assuming that this is a separate repo from SCOREC/core, rather it is a personal repo.  Can you provide some instructions for how best to check this out as a branch "under" our standard repo.  I guess I am not git-savy enough to do that since for PHASTA we keep everything under either the public or the next repo which are ultimately both "linked" on github.

**Comment by cwsmith** [comment]
```
cd /path/to/existing/core/repo
git remote add cws git@github.com:cwsmith/core.git
git fetch cws
git checkout matchedReader
```

On my system this looks like:

```
~/develop/core (develop)$ git remote -v
origin	git@github.com:SCOREC/core.git (fetch)
origin	git@github.com:SCOREC/core.git (push)
~/develop/core (develop)$ git remote add cws git@github.com:cwsmith/core.git
~/develop/core (develop)$ git fetch cws
remote: Enumerating objects: 75, done.
remote: Counting objects: 100% (75/75), done.
remote: Compressing objects: 100% (29/29), done.
remote: Total 75 (delta 53), reused 68 (delta 46), pack-reused 0
Unpacking objects: 100% (75/75), done.
From github.com:cwsmith/core
 * [new branch]        apf_sim_config          -> cws/apf_sim_config
 * [new branch]        chef                    -> cws/chef
 * [new branch]        cws/control_print_stmts -> cws/cws/control_print_stmts
 * [new branch]        dcVtx                   -> cws/dcVtx
 * [new branch]        develop                 -> cws/develop
 * [new branch]        fusion_dev              -> cws/fusion_dev
 * [new branch]        generateAdvanced        -> cws/generateAdvanced
 * [new branch]        liipbmod                -> cws/liipbmod
 * [new branch]        master                  -> cws/master
 * [new branch]        matchedReader           -> cws/matchedReader
 * [new branch]        mis-work                -> cws/mis-work
 * [new branch]        phastaChef_mesh_measure -> cws/phastaChef_mesh_measure
 * [new branch]        sim_getDgCopies         -> cws/sim_getDgCopies
 * [new branch]        ugridhacks              -> cws/ugridhacks
~/develop/core (develop)$ git checkout matchedReader
Branch 'matchedReader' set up to track remote branch 'matchedReader' from 'cws'.
Switched to a new branch 'matchedReader'
~/develop/core (matchedReader)$

**Comment by KennethEJansen** [comment]
Thanks.  I have successfully built the code on the viz nodes and repeated the 175k test.  I further  tested the following larger cases 
Parts     Mesh size      Time to Create   Memory Peak  Time to Write smb
4             8M                    65                      4x1.7GB           26
4             125M              999                      4x27GB         350
8             ""                  1012                       8x13.5GB 
(note it is called 150M but it is actually 125M according to the code).  As expected this is linear in time and memory for 4 parts so all is looking good at this point. The mesh creation does not seem to benefit from more cores/parts but maybe that is expected?

Thanks for adding the render function and output.  I did verify the 175k and 8M cases in paraview

I will have to move to Cooley to test the 1B element case since it is projected to need > 800GB.

**Comment by KennethEJansen** [comment]
Note all of the above is with NULL passed for matching.  Please let me know when you think the code is ready to test matching.  Also, it looks like you have not made the jump to the new file formats yet (e.g., your code is still not expecting a header and is reading id as the first column on crd and cnn  at least).

**Comment by KennethEJansen** [comment]
I think we are facing a tradeoff between IO bottlenecks at large process/per node and the scalability of the mds element creation.  To test the extreme of this I tried one process per node but as as you can see below, this fails on an INT_MAX limit:

jansen@cc014: /projects/TRAssembly_2/kjansen/Models/BoeingBump/matchedReaderTest/1B $ mpirun -np 8  --hostfile $COBALT_NODEFILE  /projects/TRAssembly_2/SCOREC-core/build-matchedReader/test/matchedNodeElmReader box_WMLES_1B.cnn box_WMLES_1B.crd NULL mod mesh/
numElms 1000000000 numVerts 1003003001
isMatched 0
ERROR PCU message size exceeds INT_MAX... exiting
ERROR PCU message size exceeds INT_MAX... exiting
[cc015:mpi_rank_2][error_sighandler] Caught error: Aborted (signal 6)
[cc009:mpi_rank_5][error_sighandler] Caught error: Aborted (signal 6)
ERROR PCU message size exceeds INT_MAX... exiting
[cc003:mpi_rank_6][error_sighandler] Caught error: Aborted (signal 6)
ERROR PCU message size exceeds INT_MAX... exiting
[cc010:mpi_rank_3][error_sighandler] Caught error: Aborted (signal 6)
ERROR PCU message size exceeds INT_MAX... exiting
[cc004:mpi_rank_4][error_sighandler] Caught error: Aborted (signal 6)
ERROR PCU message size exceeds INT_MAX... exiting
[cc005:mpi_rank_1][error_sighandler] Caught error: Aborted (signal 6)
ERROR PCU message size exceeds INT_MAX... exiting
[cc014:mpi_rank_0][error_sighandler] Caught error: Aborted (signal 6)

**Comment by cwsmith** [comment]
Thanks for testing.  Glad to see the tool worked for 1B elements.  I've added that result to the table.

| Parts  |   Mesh size   |   Time to Create (s) |  Memory Peak (GB) | Time to Write smb  (s) |
| --     |  --           |  --              | --            | --                 |
| 4      |       8M      |              65  |  4x1.7      | 26                 |
| 4      |       125M    |          999     |  4x27       | 350                |
| 8      |       ""      |            1012  |  8x13.5     |                    |
| 32     |       1B     |     3589  |  32x31?     |          22          |
| 64      |       ""      |      3433  |  ?     |  13                  |

I think we'll have to move to MPI_IO and binary files to get the read times down.  This would append another bullet on item 4 from [here](https://github.com/SCOREC/core/issues/205#issuecomment-451579947). Matlab supports binary file writing:
https://www.mathworks.com/help/matlab/ref/fwrite.html

Matching is not supported yet.  I'll post an update here when it is ready for testing.

**Comment by KennethEJansen** [comment]
Since this approach appears promising, I am going to take some time here to plan here the remaining steps to putting this into a production run.  I will put tentative assignments  via initials but of course we can redistribute as needed:
1) Preliminary testing for meshes up to 1B .  This is nearly complete as noted above by KEJ.  
2) Adding matching capability (CWS).
3) Switch over to the header plus removal of ID column  (CWS).
4) Adding classification capability (CWS).
5) Creation of "real" meshes (current ones are isotropic boxes) with spatially varying 2D boundary layers and vertical filling to top boundary (RB). Retest should be easy since, other than bugs, it is just coordinate changes.
6) Adding the load of flat text BC and IC description a la spj or smd (CWS or other if needed since I think there is a prospect that what we have will just work as long as 4) is done in a  way that is compatible). 
7) With 6) complete we should be able to start testing in PHASTA (RB).
8) Push this up to our larger target meshes (RB).
9) Adding Multi-topology (CWS)
10) Adding MPI_IO and binary (CWS).  This is a low priority for now.  Of course it would be nice but I think we can get to > 10B without it.  1B element case on 16 nodes with one process takes about 1/2 an hour to read the mesh.  KEJ has modified the code to measure this directly and will get data on that at 2 and 4 processes per node as time allows but since we were able to get 1 B through on 8 nodes, I suspect we can get 10B through on 80 (or probably much less) which we have in Cooley.

**Comment by cwsmith** [comment]
What is the notation in the parts column?  Is it <total number of processes>x<total number of nodes> ?

The list of todo's looks good to me.

Is 10B the upper limit for mesh size?

**Comment by rickybalin** [comment]
Yes 10B should be the upper limit.


-----------------------------------------------------------------
Riccardo Balin
Doctoral Candidate
Smead Aerospace Engineering Sciences
University of Colorado, Boulder
Riccardo.Balin@Colorado.EDU

**Comment by rickybalin** [comment]
To update on number 5 on the list of steps above, I am able to generate meshes of the desired geometry and with the desired features (it is missing a spatially varying BL height but that is easy to add).

Here is a picture of the mesh I created. Note this is only a 35k element mesh, but should be good enough to test that this workflow is successful with a simple RANS simulation. 

![mesh](https://user-images.githubusercontent.com/10522736/50849625-cc666300-1334-11e9-9f85-9fb3fe991ab1.png)

**Comment by KennethEJansen** [comment]
This looks good.   I am going to add to task 5 what you probably already know goes with it -- modifying your code to write the model classification  and boundary condition files which are needed by task 6.

**Comment by rickybalin** [comment]
I started working on that, but could not find a rule of thumb for the order of numbering of the geometric entities (faces, edges and vertices). I am making up one, but will be able to modify the code with another order easily.

**Comment by KennethEJansen** [comment]
Unless Cameron says otherwise, I suggest you follow the order of the connectivity array. 
Model vertices start with 0 at the "origin" which we can define as minx, miny, minz and then there are three other model vertices which cycle in an order that their curl points into the domain.  These 4 vertices define the first four edges as well (e0 = v1-v0, e1= v2-v1, etc).  These first 4 edges define face 0 (f0).   The other 4 model vertices follow the same order but shifted into the third dimension (e.g v4 is a third dimension displacement from v0 and others shift this relationship by 1 so that v7 is extension of v3).  The next 4 edges are numbered such that their origin is  v{0..3} and their endpoint is v{4...7} (e.g., they are directed into the third dimension (e.g., e4=v4-v1)).   The last 4 edges are "parallel" to the first 4 and form the loop for the face opposite face 0 (e.g., e8=v5-v4).  The next 4 faces are formed such that their "base" is on the first 4 edges (e.g., f1 has a base with e0 and in general f{1..4} has a base with edge{0..3}.  Finally, face 5 closes off the box opposite of f0.

**Comment by rickybalin** [comment]
Code for the classification file is also done, and I gave the file a ".class" extension. This file has the mesh node global ID in the first column, and the classification tag in the second column, and no title since it seems like the title is not wanted in the other files.

For the BC and IC, I simply copied the .spj file that had the desired boundary and initial conditions for a .smd model with the same geometry, and modified the face and model tags to match what was done in the classification file. The tag I used for what Simmetrix and SimModeler call the region is 0 because we said to give a classification of 0 to the fully interior nodes, but I can change that if 0 is not an appropriate value.

This should complete the set of files required, which for the hill geometry can be found on the viz nodes under the name "2DGaussHill_test" at: /users/rbalin/NSF_DataDrivenTurbMod/PrelimTests/matchedNodeElmReader/2Dhill/test

**Comment by cwsmith** [comment]
@rickybalin Looking back at the `box_small` case there appears to be problem with the matching file.  Specifically, if vtx 1 is matched with vtx 51, as indicated by line one, 
```
$ head -n 1 box_small.match 
1 51
```
I'd expect to see that vtx 51 was matched with vtx 1 on line 51.  Instead it looks like vtx 51 is not matched.
```
$ sed '51q;d' box_small.match 
51 -1
```

**Comment by KennethEJansen** [comment]
Cameron, if it makes your matching work easier, It would be easy for us to restrict ourselves to cases where there was only on-part matching.   

This is as easy as being sure to choose 
numVerts/numParts=k*Nz
where k is strictly an integer, Nz is the number of points in the matched direction (z in this case), numVerts is the the total number of vertices and numParts is the number of parts we will steam this into.  

Since numVerts=Nx*Ny*Nz, this is equivalent to
Nx*Ny=numParts*k

which also says there are k, Nz length lines of nodes on each part. 

This is  a super easy constraint for us to live with if it makes your life easier.

**Comment by KennethEJansen** [comment]
We should also plan for which stage we will switch the file formats (e.g., add the header with the array shape and ditch the ID column) .

**Comment by cwsmith** [comment]
See #211 for progress on matching.

Moving @KennethEJansen posts on classification here:

> vertex: whatever it is classified on
> 
> edge classification: An edge in the mesh will have 2 vertices. From their classification the following classification pairs are possible and those pairs -> edge_classification
> v v ->e
> v e ->e
> v f -> f
> v r -> r
> e e -> e
> e f -> f
> e r -> r
> f f -> f
> f r -> r
> r r -> r
> 
> face classification:
> if any of four vertices are classified on region -> region
> else on model face and it is impossible to have more than one face in the 4 vertices classification
> 
> region -> region

**Comment by KennethEJansen** [comment]
Since, in the above, there is a general rule that classification is always set by the node with the highest dimensional closure, I think you can encapsulate all the above with
entClass=max(nodesOfClosureClass)
or loop over the nodes of the entity and assign the max of the nodes to the entity

Because we have numbered higher dimensional model entities with higher number this works.  Initially I thought it would be good to change our model entity numbering to be 10^d+entNumberD where d is the  dimension of the entity and entNumberD was counting of entities of dimension D but, since we always went to higher numbers when we started numbering entities what we had already is sufficient.

**Comment by rickybalin** [comment]
Task list for Classification:
1) Set up model adjacency (what model faces close model region,  what model edges close each model face, ...). Do we have to do this? We might have to do this if we are going to use inheritance of boundary conditions.
2) For each entity in the mesh, set model classification.
    a) every mesh region classified on a model region
    b) every mesh face classified on a model face or model region
    c) ...

Ken and I made some progress in /users/rbalin/matchedReader/core/test/matchedNodeElmReader.cc. We have a version that compiles, but crashes during the call to apf::setCoords(mesh, m.coords, m.localNumVerts, outMap), before any call to classification related code. It does not seem to be something we broke adding our modifications for classification, but I am still in the process of understanding why it happens.

Ken, I had to modify our initial work because of how setModelEntity() is defined, but it should do the same thing.

**Comment by cwsmith** [comment]
Can you confirm that commit https://github.com/cwsmith/core/commit/75db2593d0d495b96ca0b3148c7d98aa01c75295 builds and runs?  The code I added for reading classification etc. may be at fault.

**Comment by rickybalin** [comment]
It builds, but does not run. Stops at the same point in apf::setCoords, line 224.

**Comment by cwsmith** [comment]
OK.  Thank you for checking.  I should be able to debug this for about an hour or so tomorrow.

**Comment by rickybalin** [comment]
Thank you, please let us know if there is more we can do to help.

**Comment by KennethEJansen** [comment]
What broke that version was that, when you were testing 2D you added the reading of three lines at the top of the .cnn file 
1 1
<nsd>
<numel> <nvtxPerElement>

adding these three lines  to our cnn file gets 
https://github.com/cwsmith/core/commit/75db2593d0d495b96ca0b3148c7d98aa01c75295
to run for the through (have not checked the result).

**Comment by cwsmith** [comment]
Ahh yes.  Thank you.  I added that to deal with 2d meshes and multiple topologies.

**Comment by rickybalin** [comment]
Is there a way to get the coordinates of a MeshEntity when that entity is a mesh vertex? And is there a way to get the coordinates of Adjacent vertices that come out of getAdjacent when it is called on a mesh face or edge (i.e. apf::MeshEntity* e; apf::Adjacent verts; mesh->getAdjacent(e,0,verts);)? 

I tried looking in https://scorec.rpi.edu/~seol/scorec/doxygen/apfMesh_8h.html but couldn't see something useful.

Thanks

**Comment by hapfang** [comment]
Not sure if it is directly applicable in your case, but here is an example to access the vertex coordinates: 
https://github.com/PHASTA/core/blob/3d5071a1af31c65a0c81b27af00e955472b11050/phasta/phAdapt.cc#L158-L170

**Comment by rickybalin** [comment]
Thanks Jun, that helped!

**Comment by KennethEJansen** [comment]
Three points:   
1) Above,  I have fixed mistakes about how the model classification  of the set of vertices  for a given mesh entity control what model that entity should be classified on.  I had it backwards the first time I wrote it when I said it would be the max(vertex_classification) because I forgot we numbered highest dimensional entities first. I think I have replaced all the times I said "max" with "min". The code that Riccardo and wrote yesterday uses this min characteristic.
2) Above we made a task list to get classification working.  In the first task we left as a question whether it was necessary to actually create all the model entities (e.g., 6 model faces, 12 model edges, and 8 model vertices) and add them to the model which, as I understand it, only has the notion of a model region at the start?  Further, do we need to also define the topology of that model?  As noted above I think we will need this if we expect inheritance of BC's to work (e.g., typically we set BCs only on model faces but expect entities classified on the model edge to inherit or boolean the two model face BC's to the model edge or more specifically, to know what BC to place on mesh entities classified on model edges).  Assuming the answer is yes, Cameron, can you point us to any example code that adds model faces, edges, and vertices to a gmi model?  Further, can you point to examples that have defined the model topological relation between the model entities (e.g., all the model faces that create the shell of a model region, the model edges which form a loop around a model face, and finally, the model vertices which define the start and end of a model edge. If that is HARD, could we get away with not using inheritance and listing every BC on every model entity? 
3) Is it really necessary for us to classify mesh edges and mesh regions?  Put another way, what happens if we don't under the assumption that we are never going to probe that classification?  The only time we would probe edges is if we were using the hierarchical basis which puts DOF's of mesh edges. AFAIK, our IC's only are on vertices and don't even use classification at all.  Essential BC's are assigned to mesh vertices only in this case.  Natural boundary conditions are probably built up from iterators over mesh faces and do use their classification on model faces.  Region iterators do exist to find connectivity but this does not involve classification.

To be clear, what Riccardo and I have written thus far just does the following:
1) find the min vtx classification for the verts of an entity by pulling them out of  getIntTag(v,vtxClass,&c)
2) based on the min value of "c" across the vertex list AND the value of c we know what dimension that entity should be classified on e.g.,   


`  if (c == INTERIORTAG) {
       mesh->setModelEntity(v,getMdlRgn(model));
    } else if (c >= FACE && c <= FACE_LAST) {
       mesh->setModelEntity(v,getMdlFace(mesh,c));
    } else if (c >= EDGE && c <= EDGE_LAST) {
       mesh->setModelEntity(v,getMdlEdge(mesh,c));
    } else if (c >= VERTEX && c <= VERTEX_LAST) {
       mesh->setModelEntity(v,getMdlVtx(mesh,c));
    }`

BUT I am unclear if this does the write thing if those model faces, edges, and vertices were not only created but also have model adjacency/topology prior to this classification (or after???).

**Comment by cwsmith** [comment]
@KennethEJansen 
2a. No geometric model entities exist prior to the assignment of classification.  AFAIK, you can get through mesh verification with all mesh entities classified on a single region.  Likewise, I think the number of geometric model entities classified against is up to the user.
2b. I'll look into this.  I have not created the topology of a geometric model before.  I don't see how inheritance would work without the geometric model topology defined.
3. Yes; all mesh entities need to have classification set.

**Comment by cwsmith** [comment]
This looks like it could be helpful:

https://github.com/SCOREC/core/blob/7ba0ce41fee3369b8fa8f0233799552ab721d5d4/mds/apfBox.h

https://github.com/SCOREC/core/blob/master/mds/apfBox.cc

the user facing function

https://github.com/SCOREC/core/blob/7ba0ce41fee3369b8fa8f0233799552ab721d5d4/mds/apfBox.h#L64-L74

 wants to create the mesh and the box model (with topology).  We could probably change this to just give us the model.

**Comment by cwsmith** [comment]
If we could get this to produce a box model (possibly via a modified [buildModel()](https://github.com/SCOREC/core/blob/7ba0ce41fee3369b8fa8f0233799552ab721d5d4/mds/apfBox.cc#L135) ) then we want to use that call instead of creating a null model here:

https://github.com/cwsmith/core/blob/d0132907a9bd96263167551c39285c720c97b25c/test/matchedNodeElmReader.cc#L346

**Comment by cwsmith** [comment]
Yes, we need to remove the `deriveMdsModel(...)` call if we are using `buildModel(...)`.

Suppose we have created a model with `buildModel(...)` and we know the integers associated with each geometric model entity in that model (via the changes to `formModelTable(...)`, IIUC).  Then, knowing the map of the integers that have been attached via an `apf::MeshTag` to the geometric model entities ([the `vtxClass` tag](https://github.com/cwsmith/core/commit/d0132907a9bd96263167551c39285c720c97b25c#diff-46221101485692e5baf9643d8cf3e8baR67) ), we can call `mdlEnt = gmi_find(..., <geometric model id>)` and `mesh->setModelEntity(meshEnt,mdlEnt)` to set classification.

**Comment by cwsmith** [comment]
I think we'll have to make a version of the BoxBuilder constructor:

https://github.com/SCOREC/core/blob/7ba0ce41fee3369b8fa8f0233799552ab721d5d4/mds/apfBox.cc#L68-L84

without these lines:

https://github.com/SCOREC/core/blob/7ba0ce41fee3369b8fa8f0233799552ab721d5d4/mds/apfBox.cc#L81-L83

**Comment by KennethEJansen** [comment]
Yep that was the first thing I tried.  My C++ is too weak though and I hit a wall.  The example box.cc calls 

 69   apf::Mesh2* m = apf::makeMdsBox(nx,ny,nz,wx,wy,wz,is);
 70   gmi_model* g = m->getModel();

That is , it gets the model from the mesh and in our case we are doing the mesh from file AFTER the model is created.  That is why I was trying to get at the interior functions that don't depend on the mesh.

**Comment by cwsmith** [comment]
How about running the 'box' test driver to generate a .dmg model of the necessary dimensions, throwing away the mesh that is also produced, and then changing the model entity numbering scheme written by @rickybalin's mesh generation script to match the model entity numbering from apfBox.cc ?

To build the 'box' driver:

`make -j box`

then run it with:

`./test/box`

to see the input args.

**Comment by KennethEJansen** [comment]
We could try this.  I have already run box in the debugger to understand how it  "worked".  A slight mod to what you are saying is to make the one line change to box so that it starts its tag numbers from 31, 11, 1, and 0 so that we actually get the same tag ranges.  This is important because we are using these tag ranges to do the classification.  I will try this quickly.

Can you also comment on whether there is any difference between a tag numer and a model entity id in the code you started and we took over (e.g., lines 22 to 32 ish)?

**Comment by cwsmith** [comment]
OK.  Sounds good.

> Can you also comment on whether there is any difference between a tag numer and a model entity id in the code you started and we took over (e.g., lines 22 to 32 ish)?

They are just integer ids.  If the ints attached to the mesh vertices are exactly the same as the model entity ids then the lines 

https://github.com/cwsmith/core/blob/d0132907a9bd96263167551c39285c720c97b25c/test/matchedNodeElmReader.cc#L21-L31

are not needed.

**Comment by KennethEJansen** [comment]
This built and is failing in verify now so I think that means we still have work to do with classification.  Riccardo, I tested this in your VNC
 2582  mpirun -np 1 /projects/tools/SCOREC-core/btest/test/box 5 5 5 1 2 3 0 modbox.dmg meshbox/
 2583  mpirun -np 2 /projects/tools/SCOREC-core/btest/test/matchedNodeElmReader  box_small.cnn  box_small.crd NULL box_small.class model mesh/
rbalin@viz003: ~/NSF_DataDrivenTurbMod/PrelimTests/matchedNodeElmReader/box/kejTest

in a subdirectory from where you were testing.   The output suggests we have the classification wrong, perhaps simply because the faces and edges dictated by my hack of box, don't match your file e.g.,
numVerts 27
isMatched 0
seconds to create mesh 0.002
APF FAILED: apf::Verify: vertex with 3 adjacent edges
centroid: (0, 0, 0)
based on the following:
 - vertex is classified on a model region
we would expect the adjacent edge count to be at least 4

[viz003:29720] *** Process received signal ***
APF FAILED: apf::Verify: vertex with 3 adjacent edges
centroid: (1, 0, 0)
based on the following:
 - vertex is classified on a model region
we would expect the adjacent edge count to be at least 4

though that may mean that something that should not be classified on a region is getting classified there.

**Comment by KennethEJansen** [comment]
I am in the debugger and finding that  before I execute line 182 I have what looks like valid information in a totalview dive on model.   After I step past line 182 I get "bad address" when I dive on model. I guess that means that something in the getMdlRgn function is corrupting the model?

181 void setRgnClassification(gmi_model* model, apf::Mesh2* mesh) {
182   apf::ModelEntity* mdlRgn = getMdlRgn(model);
183   apf::MeshIterator* it = mesh->begin(3);
184   apf::MeshEntity* rgn;
185   while( (rgn = mesh->iterate(it)) )
186     mesh->setModelEntity(rgn,mdlRgn);
187   mesh->end(it);
188 }
189 
190 void setClassification(gmi_model* model, apf::Mesh2* mesh, apf::MeshTag* t) {
191   setRgnClassification(model,mesh);
192   setFaceClassification(model,mesh,t);
193   setEdgeClassification(model,mesh,t);
194   setVtxClassification(model,mesh,t);
195   mesh->acceptChanges();
196 }

I wonder if we are missing something significant still.  I did comment out the call to derive model BUT don't we have to do SOMETHING to associate the model with the mesh? In rough terms it seem we are:
1) reading the model
2) reading the coordinates and connectivity
3) using apf:construct to build the mesh in the mds form
4) setting the coordinates
5) applying classiification

Is it step 5) that does the model/mesh association?

**Comment by cwsmith** [comment]
I assume that the `model` object that is returned by `gmi_load("/path/to/the/dmg/from/box")` is being passed to the create empty mesh call?

https://github.com/cwsmith/core/blob/d0132907a9bd96263167551c39285c720c97b25c/test/matchedNodeElmReader.cc#L347

**Comment by KennethEJansen** [comment]
Probably should have created a new comment since I am not sure updates go out.  Issue is line 36 and 37.  Here is what I did for the model load:

 443  gmi_model* model = gmi_load("modbox.dmg");


there are no complaints about this in the following code
452   apf::Mesh2* mesh = apf::makeEmptyMdsMesh(model, m.dim, isMatched);
453   apf::GlobalToVert outMap;
454   apf::construct(mesh, m.elements, m.localNumElms, m.elementType, outMap);
455   delete [] m.elements;
456   apf::alignMdsRemotes(mesh);
457 //  apf::deriveMdsModel(mesh);
458   /*for (int i=0; i<81; i++) {
459   std::cout<<m.coords[i]<<std::endl;
460   }*/
461   apf::setCoords(mesh, m.coords, m.localNumVerts, outMap);
462   delete [] m.coords;
463   if( isMatched ) {
464     apf::setMatches(mesh, m.matches, m.localNumVerts, outMap);
465     mesh->acceptChanges();
466     delete [] m.matches;
467   }
468   apf::MeshTag* t = setIntTag(mesh, m.classification, 1,
469       m.localNumVerts, outMap);
470   outMap.clear();
471   setClassification(model,mesh,t);

but within setClassification, the corruption on line 36 and 37 is what I am concerned about

**Comment by rickybalin** [comment]
Sorry for the late contribution, I was working on something last night but wasn't sure about it and some of this morning's discussion may have helped me. I could still be missing a major point that makes my approach flawed, but I think with some hard coding in the classification routines I might be able to get something that works.

I changed makeMdsBox to output a model instead of a mesh, so in matchedNodeElmReader.cc I replaced line
gmi_model* model = gmi_load(".null");
with
gmi_model* model = apf::makeMdsBox(2,2,2,1,1,1,0);

I currently have to comment out the call to setClassification for this code to work, but with this temporary comment it appears to work fine. I am able to see the mesh I intended to make in Paraview, and the dmg model I create also has the right format. Of course, the numbering of the model vertices, edges and faces is different from what our tags are, so my classification code was not working because all the getMdlRgn, getMdlFace, ecc functions were looking for model entities with ids that do not exist in the model. So if I can relate the ids in the dmg model to the tags that I print out in the .class file, I think this approach might work.

**Comment by KennethEJansen** [comment]
The issue with line 36 and 37 may be a false alarm.  Totalview says bad address but perhaps it is just that it can't figure out what we are doing there.  We do get a result from the function and and  iterator from that.   

Riccardo's way also is working we think but ONLY if we use the zero based tag numbering.  The simple shift (only change) causes verify of the mesh to crash (makes no sense because it happens even if we skip classification)??

**Comment by KennethEJansen** [comment]
Note that what Riccardo did was to make apfBox.cc return a model, not a mesh so obviously this breaks other uses of apfBox.cc.  I think it makes more sense to modify apfBox to publicly expose the ability to make just a model, just a mesh, or both so that we can have code reuse instead of code duplication.
It would also be good to understand why we cannot use tags that start their count away from zero?  I am guessing somewhere in the code it is hard coded to assume that it starts from 0?

**Comment by rickybalin** [comment]
I was able to use this tool to create the mesh for the bump geometry we are interested in shown below, successfully apply initial and boundary conditions with Chef, partition the mesh and run PHASTA with a RANS turbulence model to get the velocity field shown below as well.

The only thing to note is that to achieve this, I had to change the sign of the assert in checkBoundaryVertex (in phOutput.cc) for the direction of the normal vectors for hex elements.

Thanks a lot for everyone's help.

![image](https://user-images.githubusercontent.com/10522736/52091357-622a9200-2571-11e9-9bd0-05fb6912cb73.png)

![image](https://user-images.githubusercontent.com/10522736/52091390-7f5f6080-2571-11e9-89bd-50640737c71b.png)

**Comment by cwsmith** [comment]
If the code to set classification is stable would you please push the changes to the branch? I sent you both invites to my fork so you can push.

**Comment by rickybalin** [comment]
I can certainly do that, but to make matchedNodeElmReader.cc work I made some changes to makeMdsBox that would brake other routines that use it. Should I still push?

Also, classification was testes successfully in parallel.

**Comment by cwsmith** [comment]
Can you test if using the 'box' driver would create a model that works with `matchedNodeElmReader`? If so, we don't have to change any of the makeMdsBox code.  If you guys would really prefer to have this all in one executable then we could 1) based on the changes you made, create an additional makeMdsBox API that only creates the model or 2) throw away the small mesh produced by the original makeMdsBox for essentially zero cost.

**Comment by cwsmith** [comment]
@KennethEJansen wrote:

> I have committed two branches to your repo.
> withClass was our attempt to read the model created by the box executable but this crashed
> withClassWorks is the code where Riccardo creates the model through a modification to the function the box code uses to return a model instead.
> 
> Note the latter has additional code to re-map the model entity numbering that box created to get the correct result but I don’t think this is the cause for the first approach to fail.

These branches are in SCOREC/core:
https://github.com/SCOREC/core/tree/withClass
https://github.com/SCOREC/core/tree/withClassWorks

**Comment by rickybalin** [comment]
Hi Cameron,

We are trying to add another MeshTag in the same way that classification is added, so we copied the line 
apf::MeshTag* t = setIntTag(mesh, m.classification, 1, m.localNumVerts, outMap);
and changed it to
apf::MeshTag* tf = setIntTag(mesh, m.fathers2D, 1, m.localNumVerts, outMap);
where m.fathers2D contains the data we need to perform spanwise averaging in PHASTA.

I am having problems adding the second tag though. Depending on where I place the second call to setIntTag I get different errors, but I have not yet been able to make it work. Do we have to create a new IntTag with createIntTag before we can add the second one?

Currently I have this:
  ```
apf::MeshTag* tc = setIntTag(mesh, m.classification, 1,
      m.localNumVerts, outMap);
  outMap.clear();
  setClassification(model,mesh,tc);
  apf::removeTagFromDimension(mesh, tc, 0);
  mesh->destroyTag(tc);
 
  apf::MeshTag* tf = setIntTag(mesh, m.fathers2D, 1,
      m.localNumVerts, outMap);
  apf::removeTagFromDimension(mesh, tf, 0);
  mesh->destroyTag(tf);

  if(!PCU_Comm_Self())
    fprintf(stderr, "seconds to create mesh %.3f\n", PCU_Time()-t0);
  mesh->verify();

  gmi_write_dmg(model, argv[6]);
  mesh->writeNative(argv[7]);
  apf::writeVtkFiles("rendered",mesh);

  mesh->destroyNative();
  apf::destroyMesh(mesh);
  PCU_Comm_Free();
  MPI_Finalize();
```
Interestingly, if I comment out all the classification related code (so the first block), the executable runs successfully with the fathers2D tag code still active. So it seems like my problem is with calling setIntTag twice specifically.

Any advice on this is very appreciated.

Thanks.

**Comment by cwsmith** [comment]
I just pushed a 'fix':

https://github.com/SCOREC/core/commit/3a94fa67345324c9d3559bc3b29d23eed435bcdc

**Comment by rickybalin** [comment]
Thanks a lot, those changes fixed the crash.

**Comment by rickybalin** [comment]
I just pushed to cwsmith/core in the branch addNfath a working version of matchedNodeElmReader and chef that create the data structures needed for spanwise averaging in PHASTA and write them to the geombc files.

**Comment by KennethEJansen** [comment]
@cwsmith  I have prepared a test case for this code that is of the same geometry BUT is all tet instead of all hex. As I see it, the crd, match, and class files are nodal data structures and are thus unaffected by how the nodes are connected into elements.  Can you comment on whether there was an assumption at any point in the software that the elements would be hexes or that the faces on the boundary would be quads?

**Comment by KennethEJansen** [comment]
Here is the output from my first attempt:
kjansen@viz003: /projects/tools/Models/BoeingBump/LES_DNS_Meshing/MGEN-Mesh $ mpirun -np 8 /users/rbalin/matchedReader/build-addNfathDebug/test/matchedNodeElmReader geom3D.cnn ../LES_DNS_ReL1M_3d.crd ../LES_DNS_ReL1M_3d.match ../LES_DNS_ReL1M_3d.class ../LES_DNS_ReL1M_3d.fathers2D t1.dmg t1.smb 
numVerts 95719374
0 readMatches numvtx 95719374 localnumvtx 11964921 firstVtx 0 lastVtx 11964921
1 readMatches numvtx 95719374 localnumvtx 11964921 firstVtx 11964921 lastVtx 23929842
2 readMatches numvtx 95719374 localnumvtx 11964921 firstVtx 23929842 lastVtx 35894763
5 readMatches numvtx 95719374 localnumvtx 11964921 firstVtx 59824605 lastVtx 71789526
7 readMatches numvtx 95719374 localnumvtx 11964927 firstVtx 83754447 lastVtx 95719374
3 readMatches numvtx 95719374 localnumvtx 11964921 firstVtx 35894763 lastVtx 47859684
4 readMatches numvtx 95719374 localnumvtx 11964921 firstVtx 47859684 lastVtx 59824605
6 readMatches numvtx 95719374 localnumvtx 11964921 firstVtx 71789526 lastVtx 83754447
isMatched 1
seconds to create mesh 2278.281
  - verifying tags: fathers2D
APF FAILED: apf::Verify: triangle with 2 adjacent regions
centroid: (0.914777, 0.00028926, 6.25978e-05)

I am willing to accept that I may have something wrong in my mesh BUT, from the text and my knowledge of the geometry (xmax=9.149659131893401E-01, zmin=0, ymin=2.851842731373213E-13), I am fairly certain that this triangle has one vertex classified on the "corner: model edge and the other two on the  two faces (xmax and zmin) that intersect there.  My guess is that when we only provided node classification we did not anticipate that  in this scenario, the face SHOULD be classified on the interior.  From the message it seems it was not.

Does this make sense?

**Comment by cwsmith** [comment]
@KennethEJansen 
1) Mono-topology meshes should work with this version.
2) Without a deeper dive into this thread and the code, I'm guessing the logic for processing vertex-only classification data for this geometry was hardcoded for hexes.  The discussion of classification logic appears to start here:
https://github.com/SCOREC/core/issues/205#issuecomment-455326494

**Comment by KennethEJansen** [comment]
yep.   I wrote that and it is wrong.  I will dig on the code and see if I can fix it.

**Comment by KennethEJansen** [comment]
`diff --git a/test/matchedNodeElmReader.cc b/test/matchedNodeElmReader.cc
index 442a4d2c..3dacd39a 100644
--- a/test/matchedNodeElmReader.cc
+++ b/test/matchedNodeElmReader.cc
@@ -264,11 +264,13 @@ void setFaceClassification(gmi_model* model, apf::Mesh2* mesh, apf::MeshTag* vtx
   while( (f = mesh->iterate(it)) ) {
     mesh->getAdjacent(f, 0, verts);
     int cmin=100;
+    int cmax=-100;
     for(size_t i=0; i<verts.size(); i++) {
       mesh->getIntTag(verts[i],vtxClass,&c);
       cmin=std::min(cmin,c);
+      cmax=std::max(cmax,c);
     }
-    if (cmin == INTERIORTAG) {
+    if (cmin == INTERIORTAG || cmin != cmax) {
        mesh->setModelEntity(f,getMdlRgn(model));
        //cint++;
     } else if (cmin >= FACE && cmin <= FACE_LAST) {
`

**Comment by KennethEJansen** [comment]
```diff@@ -633,8 +635,8 @@ int main(int argc, char** argv)
   apf::removeTagFromDimension(mesh, tc, 0);
   mesh->destroyTag(tc);
  
-  apf::MeshTag* tf = setIntTag(mesh, "fathers2D", m.fathers2D, 1,
-      m.localNumVerts, outMap);
+  // not used apf::MeshTag* tf = setIntTag(mesh, "fathers2D", m.fathers2D, 1,
+// not used       m.localNumVerts, outMap);
   //mesh->destroyTag(tf);
 
   /* // Print the father2D tags
```

when I went to compile this code, apparently my flags are set such that warnings are promoted to errors and thus I had to comment the two lines that  have // not used in them because tf is not used.  That said, I see in Riccardo's git comments that he felt he DID have to setIntTag for fathers2D.  Cameron can you comment on whether setIntTag does do something that might be needed later even if tf is not used in this routine?  If yes what is the best way to handle this.  I guess I can change the compile env such that warnings are not errors or I guess I can make some trivial fake usage of tf so that the warning goes way (e.g., fprintf it to /dev/null) but this seems odd to me that compilers are pushing to such extremes.  Is a thrd alternative to replace tf  by NULL or something similar?

**Comment by cwsmith** [comment]
`(void) tf;` should work as a no-op.

**Comment by KennethEJansen** [comment]
Thanks.   That said, I was still naive about all the ways a tri face can get classified.  I think I now have all the cases covered but it, in the end, the worst case required a hard code because the proverbial corner case, where the triangle has one vertex classified on the model vert in the corner and one vertex classified on each of the model edges incident on that corner, NONE of these nodes has knowledge of which model face THIS face should be classified on.  In this sense, nodal classification appears to be insufficient to classify tri faces of tets against even a model as simple as a box.  Longer term, I think we will have to provide surface element lists with face classification flag for each of those faces but for now I have hard coded it to use the two edge and vertex classification to determine the face it belongs to.

**Comment by KennethEJansen** [comment]
pasting my diff attempt to fix chef for unstructured grid nsons array.  Riccardo please review if you can:
``diff --git a/phasta/phOutput.cc b/phasta/phOutput.cc
index 0e527be8..7bb0fb3a 100644
--- a/phasta/phOutput.cc
+++ b/phasta/phOutput.cc
@@ -944,8 +944,8 @@ static void getEdges(Output& o, apf::Numbering* vn, apf::Numbering* rn, BCs& bcs
 static void getSpanwiseAverageArrays(Input& in, Output& o) {
   if (in.spanAvg == 1) {
     apf::Mesh* m = o.mesh;
-    gmi_model* gm = m->getModel();
-    int nnodes = m->count(0); // number of nodes of wole mesh or part??
+// not used    gmi_model* gm = m->getModel();
+    int nnodes = m->count(0); // number of nodes of whole mesh or part??
     /* this will come from the adapt.inp file and is constant for all geombc
      it is the total number of father nodes, nx*ny, and each geombc loads this */
     int nfather = in.nfathers;
@@ -958,7 +958,8 @@ static void getSpanwiseAverageArrays(Input& in, Output& o) {
     o.arrays.nsonsArr = new int[nfather]; //initialize nsonsArr
     for (int i=0; i<nfather; i++) { // fill nsonsArr
       /* set each entry in nsonsArr[nfather] to equal nsons */
-      o.arrays.nsonsArr[i] = nsons; // this is point2nsons(nfath) in PHASTA
+      o.arrays.nsonsArr[i] = 0; // not alwasy structured nsons; // this is point2nsons(nfath) in PHASTA
+                           // -1 if father is not counted in sons .... I think it is so leaving it at 0
     }
     apf::MeshEntity* v;
     apf::MeshIterator* it = m->begin(0);
@@ -974,6 +975,7 @@ static void getSpanwiseAverageArrays(Input& in, Output& o) {
     while ((v = m->iterate(it))) { // loop over mesh vertices
       m->getIntTag(v,t,&tagNum);
       o.arrays.ifather[count] = tagNum;
+      o.arrays.nsonArr[tagnum]++;  // increment the nsons counter 
       //std::cout<<"Tag number "<<tagNum<<std::endl;
       count++;
     }
```

**Comment by KennethEJansen** [comment]
At the time when we are calling _setFaceClassification_ is the model adjacency known?

If yes, how do I ask for the model edges that are in the closure of a given model face?

Similarly, how do I ask for the model vertices that are in the closure (end points) of a given model edge?   I think this is the better way to go about finding higher entity classification for tets/tris.

**Comment by cwsmith** [comment]
Would you please point me at the branch on github with the current code?  

It appears a `.dmg` is loaded from previous comments in this issue.  If that is still the case, was the model created with the modified `box` driver/test?

**Comment by KennethEJansen** [comment]
I have not committed back because it is broken but I am working with the brancj referenced above (addNfath that I think is in your repo last discussed above on Feb 21 2019) https://github.com/SCOREC/core/issues/205#issuecomment-466196429

**Comment by KennethEJansen** [comment]
`624   //gmi_model* model = gmi_load(".null");
625   gmi_model* model = apf::makeMdsBox(2,2,2,1,1,1,0);
626   apf::Mesh2* mesh = apf::makeEmptyMdsMesh(model, m.dim, isMatched);
627   apf::GlobalToVert outMap;
`

What I need is the function that is equivalent to this
mesh->getAdjacent(f, 0, e);
mesh->getAdjacent(e, 0, v);

but for a model instead of a mesh.  That is, I want to get the list of model edges that close a model face and then the list of model vertices that close a given model edge.  From this it will be easy to check if the three mesh edges are in the closure of a given model face.  

That said it would also be useful to know the reverse (save code).  That is for a given model vertex, return the list of model faces that USE it.  Same for model edges, the two faces that use it.

**Comment by KennethEJansen** [comment]
I created my own adjacency table and I think I now have it getting past verify so I pushed it so that I can pull it to Cooley since I need more memory (at least I think the crash is due to running out of memory).
